obelisk/python/obelisk_py/core/control.py

```diff
@@ -0,0 +1,58 @@
+from abc import ABC, abstractmethod
+
+from obelisk_py.core.node import ObeliskNode
+from obelisk_py.core.obelisk_typing import ObeliskControlMsg, ObeliskEstimatorMsg
+
+
+class ObeliskController(ABC, ObeliskNode):
+    """Abstract Obelisk controller node.
+
+    Obelisk controllers are stateful. That is, all the quantities required to compute the control signal are stored in
+    the controller object itself. This is done because the processes generating each of these quantities may act
+    asynchronously. Similarly, the control action may be queried asynchronously.
+
+    When implementing a new ObeliskController, the user should declare all quantities required to compute the control
+    input in on_configure. These quantities should be updated by various update_X methods. Finally, the
+    compute_control method should be implemented to compute the control signal using the updated quantities. Note that
+    the control message should be of type ObeliskControlMsg to be compatible with the Obelisk ecosystem.
+    """
+
+    def __init__(self, node_name: str) -> None:
+        """Initialize the Obelisk controller."""
+        super().__init__(node_name)
+        self.register_obk_timer(
+            "timer_ctrl_setting",
+            self.compute_control,
+            key="timer_ctrl",
```

Comment:
For every key that we make, we should also provide a member variable that holds the key. Then, when the user wants to access the component, they can do it through the member variable rather than a hard-coded string. See for example [here](https://github.com/Caltech-AMBER/obelisk/blob/f08a41814d9770f1129363b6e053bb1b3f278962/obelisk/cpp/obelisk_cpp/include/obelisk_controller.h#L97).

Reply:
Resolved - but, I did this in a more general way on the python side. The key name is set as a node attribute and an exception is raised if another key with that name exists already (note that this means that even if the component type is different, the key MUST be different!)

See these lines: [link](https://github.com/Caltech-AMBER/obelisk/blob/08fed086bd71b502edecbe83051cb99b346a29ab/obelisk/python/obelisk_py/core/node.py#L404-L405)